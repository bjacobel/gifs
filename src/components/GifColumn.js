import React, { Component, PropTypes } from 'react';
import Infinite from 'react-infinite';
import { connect } from 'react-redux';

import GifWrapper from './GifWrapper';
import { getGifsAsync } from '../actions/gifs';

const mapStateToProps = (state) => {
  return {
    gifs: state.gifs,
    searchResults: state.searchResults,
  };
};

const mapDispatchToProps = {
  getGifsAsync,
};

class GifColumn extends Component {
  componentDidMount() {
    this.props.getGifsAsync();
  }

  render() {
    const { gifs, searchResults } = this.props;
    let filteredGifs = gifs;

    if (searchResults.length > 0) {
      filteredGifs = searchResults.map((result) => {
        return gifs.find(gif => result === gif.id);
      });
    }

    return (
      <Infinite
        className="gif-column"
        elementHeight={ filteredGifs.map(gif => gif.observedHeight || 400) }
        useWindowAsScrollContainer
      >
        { filteredGifs.map(gif => (
          <div className="gif" key={ gif.id + gif.src }>
            <GifWrapper gif={ gif } />
          </div>
        )) }
      </Infinite>
    );
  }
}

GifColumn.propTypes = {
  gifs: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    observedHeight: PropTypes.number,
    src: PropTypes.string,
  })),
  searchResults: PropTypes.arrayOf(PropTypes.string),
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(GifColumn);
