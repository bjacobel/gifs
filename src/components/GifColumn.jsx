import React, { Component } from 'react';
import { connect } from 'react-redux';
// import ReactList from 'react-list';

import GifWrapper from './GifWrapper';
import { getGifsAsync } from '../actions/gifs';
import { getTagsAsync } from '../actions/tags';

const mapStateToProps = (state) => {
  const { gifs, tags } = state;
  return { gifs, tags };
};

const mapDispatchToProps = {
  getGifsAsync,
  getTagsAsync
};

class GifColumn extends Component {
  componentDidMount() {
    /* eslint-disable no-shadow */
    const {
      getGifsAsync,
      getTagsAsync
    } = this.props;
    /* eslint-enable no-shadow */

    getGifsAsync();
    getTagsAsync();

    this.renderGif = this.renderGif.bind(this);
  }

  renderGif(gif) {
    const { tags } = this.props;

    return (
      <div className="gif" key={ gif.id }>
        <GifWrapper gif={ gif } tags={ tags[gif.id] || [] } />
      </div>
    );
  }

  render() {
    const { gifs } = this.props;

    return (
      <div className="gif-column">
        { gifs.slice(0, 10).map(this.renderGif) }
        {/* <ReactList
          itemRenderer={ this.renderGif }
          length={ gifs.length }
          pageSize={ 20 }
          type="variable"
        /> */}
      </div>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GifColumn);
