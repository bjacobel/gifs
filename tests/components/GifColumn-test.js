import React from 'react';
import { shallow } from 'enzyme';

import GifColumn from '../../src/components/GifColumn';

const makeGifs = (numGifs) => {
  const gifs = [];

  for (let i = 0; i < numGifs; i += 1) {
    gifs.push({
      id: i.toString(),
      observedHeight: 400,
      src: '//:0',
    });
  }

  return gifs;
};

describe('GifColumn component', () => {
  it('matches snapshot when displaying all gifs', () => {
    const gifColumn = (
      <GifColumn.WrappedComponent
        getGifsAsync={ jest.fn }
        gifs={ makeGifs(10) }
        searchResults={ [] }
      />
    );

    expect(shallow(gifColumn)).toMatchSnapshot();
  });

  it('matches snapshot when filtering by search results', () => {
    const gifColumn = (
      <GifColumn.WrappedComponent
        getGifsAsync={ jest.fn }
        gifs={ makeGifs(10) }
        searchResults={ ['1', '2', '3'] }
      />
    );

    const enzymeWrapperGifColumn = shallow(gifColumn);
    expect(enzymeWrapperGifColumn.find('Connect(GifWrapper)').length).toEqual(3);
    expect(enzymeWrapperGifColumn).toMatchSnapshot();
  });
});
