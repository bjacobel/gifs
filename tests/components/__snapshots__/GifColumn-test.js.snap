exports[`GifColumn component matches snapshot when displaying all gifs 1`] = `
<Infinite
  className="gif-column"
  displayBottomUpwards={false}
  elementHeight={
    Array [
      400,
      400,
      400,
      400,
      400,
      400,
      400,
      400,
      400,
      400,
    ]
  }
  handleScroll={[Function]}
  isInfiniteLoading={false}
  loadingSpinnerDelegate={<div />}
  onInfiniteLoad={[Function]}
  styles={Object {}}
  timeScrollStateLastsForAfterUserScrolls={150}
  useWindowAsScrollContainer={true}>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "0",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "1",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "2",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "3",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "4",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "5",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "6",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "7",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "8",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "9",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
</Infinite>
`;

exports[`GifColumn component matches snapshot when filtering by search results 1`] = `
<Infinite
  className="gif-column"
  displayBottomUpwards={false}
  elementHeight={
    Array [
      400,
      400,
      400,
    ]
  }
  handleScroll={[Function]}
  isInfiniteLoading={false}
  loadingSpinnerDelegate={<div />}
  onInfiniteLoad={[Function]}
  styles={Object {}}
  timeScrollStateLastsForAfterUserScrolls={150}
  useWindowAsScrollContainer={true}>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "1",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "2",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
  <div
    className="gif">
    <Connect(GifWrapper)
      gif={
        Object {
          "id": "3",
          "observedHeight": 400,
          "src": "//:0",
        }
      } />
  </div>
</Infinite>
`;
